### PR TITLE
[Linux/Freedesktop] Implement native file selection dialog support.

### PR DESCRIPTION
**Pull request by bruvzg**
Same as https://github.com/godotengine/godot/pull/79574 and https://github.com/godotengine/godot/pull/47499 for Linux. Should allow accessing files outside flatpak sandbox (should auto map files to `/run/user/$UID/doc/`).

- *Production edit: This partially addresses https://github.com/godotengine/godot-proposals/issues/1123.*

### CODE REVIEW

File: platform/linuxbsd/freedesktop_portal_desktop.cpp [hunk 1]
```diff
@@ -126,6 +130,313 @@ uint32_t FreeDesktopPortalDesktop::get_appearance_color_scheme() {
...
+	}
+
+	DBusMessage *reply = dbus_connection_send_with_reply_and_block(fd.connection, message, DBUS_TIMEOUT_INFINITE, &err);
+	dbus_message_unref(message);
+
+	if (!reply || dbus_error_is_set(&err)) {
+		dbus_error_free(&err);
+		dbus_bus_remove_match(fd.connection, fd.path.utf8().get_data(), &err);
+		dbus_connection_unref(fd.connection);
+		ERR_FAIL_V_MSG(FAILED, vformat("Failed to send DBus message: %s", err.message));
```
Comment (timothyqiu) [inline platform/linuxbsd/freedesktop_portal_desktop.cpp:350]:
```suggestion
		ERR_PRINT(vformat("Failed to send DBus message: %s", err.message));
		dbus_error_free(&err);
		dbus_bus_remove_match(fd.connection, fd.path.utf8().get_data(), &err);
		dbus_connection_unref(fd.connection);
		return FAILED;
```

Use `err` before freeing it. Also applies to other similar places.

File: platform/linuxbsd/freedesktop_portal_desktop.cpp [hunk 2]
```diff
@@ -126,6 +130,318 @@ uint32_t FreeDesktopPortalDesktop::get_appearance_color_scheme() {
...
+								dbus_bus_remove_match(fd.connection, fd.path.utf8().get_data(), &err);
+								dbus_error_free(&err);
+								dbus_connection_unref(fd.connection);
+								fd.connection = nullptr;
+								remove = true;
+								break;
+							}
+							dbus_message_unref(msg);
+						}
+						dbus_connection_read_write(fd.connection, 0);
```
Comment (timothyqiu) [inline platform/linuxbsd/freedesktop_portal_desktop.cpp:433]: Clearing `fd.connection` here might feed `dbus_connection_read_write` on line 433 with a null, which results in a crash.

File: platform/linuxbsd/freedesktop_portal_desktop.cpp [hunk 3]
```diff
@@ -126,6 +130,317 @@ uint32_t FreeDesktopPortalDesktop::get_appearance_color_scheme() {
...
+				DBusMessageIter var_iter;
+				dbus_message_iter_recurse(&iter, &var_iter);
+				if (strcmp(key, "uris") == 0) {
+					if (dbus_message_iter_get_arg_type(&var_iter) == DBUS_TYPE_ARRAY) {
+						DBusMessageIter uri_iter;
+						dbus_message_iter_recurse(&var_iter, &uri_iter);
+						while (dbus_message_iter_get_arg_type(&uri_iter) == DBUS_TYPE_STRING) {
+							const char *value;
+							dbus_message_iter_get_basic(&uri_iter, &value);
+							r_urls.push_back(String::utf8(value).trim_prefix("file://"));
```
Comment (timothyqiu) [inline platform/linuxbsd/freedesktop_portal_desktop.cpp:253]:
```suggestion
							r_urls.push_back(String::utf8(value).trim_prefix("file://").uri_decode());
```

e.g. Convert `%20` into whitespaces.

File: platform/linuxbsd/os_linuxbsd.cpp [hunk 4]
```diff
@@ -164,6 +164,10 @@ String OS_LinuxBSD::get_processor_name() const {
 	ERR_FAIL_V_MSG("", String("Couldn't get the CPU model name from `/proc/cpuinfo`. Returning an empty string."));
 }
 
+bool OS_LinuxBSD::is_sandboxed() const {
+	return (getenv("container") || getenv("APPIMAGE") || getenv("SNAP"));
+}
```
Comment (akien-mga) [inline platform/linuxbsd/os_linuxbsd.cpp:169]: Would it make sense to unify these checks with https://github.com/godotengine/godot/blob/eb4301b941fa211de204e37bd4d701f7e490a945/platform/linuxbsd/joypad_linux.cpp#L85-L105 so the OS method can be used there too?

Comment (bruvzg) [reply platform/linuxbsd/os_linuxbsd.cpp]: `container` env. variable seems too ambiguous, and APPIMAGE is not sandboxed, so moved the checks from the `joypad_linux` to the `OS` as is (other apps seems to use the same checks).

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**